### PR TITLE
Feature/scaffolding and build transaction

### PR DESCRIPTION
**Pull request by Sam-Jeston**
- Repo configuration
- buildTransaction feature

### CODE REVIEW

File: src/lib/transactions/build.spec.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import { expect } from 'chai'
+import { buildTransaction } from './build'
```
Comment (rhyslbw) [inline src/lib/transactions/build.spec.ts]:
```suggestion
import { Transaction } from './'
```

File: src/lib/transactions/build.spec.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import { estimateTransactionFee } from './estimate_fee'
+
+describe('buildTransaction', () => {
```
Comment (rhyslbw) [inline src/lib/transactions/build.spec.ts]:
```suggestion
describe('Transaction', () => {
```

File: src/lib/transactions/build.spec.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    const inputs = [
+      { pointer: { id: '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef', index: 1 }, value: '1000000' },
+      { pointer: { id: 'fedcba9876543210fedcba9876543210fedcba9876543210fedcba9876543210', index: 0 }, value: '5000000' }
+    ]
+
+    let outputs = [
+      { address: 'Ae2tdPwUPEZCEhYAUVU7evPfQCJjyuwM6n81x6hSjU9TBMSy2YwZEVydssL', value: '10000' }
+    ]
+
+    expect(() => buildTransaction(inputs, outputs)).to.throw(/Inputs outweigh outputs/)
```
Comment (rhyslbw) [inline src/lib/transactions/build.spec.ts]:
```suggestion
    expect(() => Transaction(inputs, outputs)).to.throw(/Inputs outweigh outputs/)
```

File: src/lib/transactions/build.spec.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    const inputs = [
+      { pointer: { id: '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef', index: 1 }, value: '1000000' },
+      { pointer: { id: 'fedcba9876543210fedcba9876543210fedcba9876543210fedcba9876543210', index: 0 }, value: '5000000' }
+    ]
+
+    let outputs = [
+      { address: 'Ae2tdPwUPEZCEhYAUVU7evPfQCJjyuwM6n81x6hSjU9TBMSy2YwZEVydssL', value: '6000000' }
+    ]
+
+    const fee = estimateTransactionFee(inputs, outputs)
```
Comment (rhyslbw) [inline src/lib/transactions/build.spec.ts]:
```suggestion
    const fee = Transaction(inputs, outputs).estimateFee()
```

File: src/lib/transactions/build.spec.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    outputs[0].value = (6000000 - Number(fee)).toString()
+    const transaction = buildTransaction(inputs, outputs)
```
Comment (rhyslbw) [inline src/lib/transactions/build.spec.ts]:
```suggestion
    const transaction = Transaction(inputs, outputs).finalize()
```

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,2 @@
+export default class CardanoJs {
```
[the remark this answers is not in the document]
Comment (Sam-Jeston) [reply src/index.ts]: Nothing! Deleted

File: src/Transaction/Transaction.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  })
+
+  return {
+    estimateFee: estimateTransactionFee.bind(null, transactionBuilder),
+    finalize: finalize.bind(null, transactionBuilder),
+    builder: transactionBuilder
+  }
+}
+
+export function finalize (
```
Comment (rhyslbw) [inline src/Transaction/Transaction.ts]: Any reason we need to export this?

File: src/Transaction/Transaction.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+export interface ClientTransaction {
+  estimateFee: (feeAlgorithm?: CardanoLinearFeeAlgorithm) => string
+  finalize: () => CardanoTransaction,
+  builder: CardanoTransactionBuilder
+}
+
+export function Transaction (inputs: TransactionInput[], outputs: TransactionOutput[]): ClientTransaction {
+  const transactionBuilder = new TransactionBuilder()
+
+  if (!inputs.length || !outputs.length) {
```
Comment (rhyslbw) [inline src/Transaction/Transaction.ts]: We could control this using a more specific type:
``` javascript
type NonEmptyArray<T> = {
    0: T
} & Array<T>
```

File: src/Transaction/Transaction.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Sam-Jeston) [reply src/Transaction/Transaction.ts]: indeed there is not

File: src/Transaction/Transaction.spec.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
+import { expect } from 'chai'
+import { Transaction } from './Transaction'
+import { TransactionOverweight, TransactionUnderweight } from './errors'
+import { TransactionInput } from './TransactionInput'
```
Comment (rhyslbw) [inline src/Transaction/Transaction.spec.ts]:
```suggestion
```

File: src/Transaction/Transaction.spec.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
+import { expect } from 'chai'
+import { Transaction } from './Transaction'
```
Comment (rhyslbw) [inline src/Transaction/Transaction.spec.ts]:
```suggestion
import { Transaction, TransactionInput, TransactionOutput } from './'
```

File: src/Transaction/Transaction.spec.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+import { TransactionOverweight, TransactionUnderweight } from './errors'
+import { TransactionInput } from './TransactionInput'
+import { TransactionOutput } from './TransactionOutput'
```
Comment (rhyslbw) [inline src/Transaction/Transaction.spec.ts]:
```suggestion
```

File: src/Transaction/TransactionInput.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
+import * as t from 'io-ts'
+
+const transactionInputPointer = t.type({
```
Comment (rhyslbw) [inline src/Transaction/TransactionInput.ts]:
```suggestion
const pointer = t.type({
```

File: src/Transaction/TransactionInput.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+  index: t.number,
+  id: t.string
+})
+
+const transactionInputAddressing = t.type({
```
Comment (rhyslbw) [inline src/Transaction/TransactionInput.ts]:
```suggestion
const addressing = t.type({
```

File: src/Transaction/TransactionInput.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+  account: t.number,
+  change: t.number,
+  index: t.number
+})
+
+export const TransactionInputCodec = t.intersection([
+  t.interface({
+    pointer: transactionInputPointer,
```
Comment (rhyslbw) [inline src/Transaction/TransactionInput.ts]:
```suggestion
    pointer,
```

File: src/Transaction/TransactionInput.ts [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+    value: t.string
+  }),
+  t.partial({ addressing: transactionInputAddressing })
```
Comment (rhyslbw) [inline src/Transaction/TransactionInput.ts]:
```suggestion
  t.partial({ addressing })
```

File: src/Transaction/TransactionOutput.ts [hunk 16]
```diff
@@ -0,0 +1,29 @@
+import * as t from 'io-ts'
+
+const transactionOutputRequiredFields = t.type({
```
Comment (rhyslbw) [inline src/Transaction/TransactionOutput.ts]:
```suggestion
const requiredFields = t.type({
```

File: src/Transaction/TransactionOutput.ts [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+  address: t.string,
+  value: t.string
+})
+
+const transactionOutputFullAddress = t.type({
```
Comment (rhyslbw) [inline src/Transaction/TransactionOutput.ts]:
```suggestion
const fullAddress = t.type({
```

File: src/Transaction/TransactionOutput.ts [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+    getCCoin: t.number
+  }),
+  cadId: t.string,
+  cadIsUsed: t.boolean,
+  account: t.number,
+  change: t.number,
+  index: t.number
+})
+
+const transactionOutputOptionalFields = t.partial({
```
Comment (rhyslbw) [inline src/Transaction/TransactionOutput.ts]:
```suggestion
const optionalFields = t.partial({
```

File: src/Transaction/TransactionOutput.ts [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+  isChange: t.boolean,
+  fullAddress: transactionOutputFullAddress
```
Comment (rhyslbw) [inline src/Transaction/TransactionOutput.ts]:
```suggestion
  fullAddress
```

File: src/Transaction/TransactionOutput.ts [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+})
+
+export const TransactionOutputCodec = t.intersection([
+  transactionOutputRequiredFields,
```
Comment (rhyslbw) [inline src/Transaction/TransactionOutput.ts]:
```suggestion
  requiredFields,
```

File: src/Transaction/TransactionOutput.ts [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+  transactionOutputOptionalFields
```
Comment (rhyslbw) [inline src/Transaction/TransactionOutput.ts]:
```suggestion
  optionalFields
```

**Pull request merged**